docs/architecture/adr-020-block-size.md

```diff
@@ -0,0 +1,65 @@
+# ADR 020: Limiting txs size inside a block
+
+## Changelog
+
+13-08-2018: Initial Draft
+
+## Context
+
+We currently use MaxTxs to reap txs from the mempool when proposing a block,
+but enforce MaxBytes when unmarshalling a block, so we could easily propose a
+block thats too large to be valid.
+
+We should just remove MaxTxs all together and stick with MaxBytes, and have a
+`mempool.ReapMaxBytes`.
+
+But we can't just reap BlockSize.MaxBytes, since MaxBytes is for the entire block,
+not for the txs inside the block. There's extra amino overhead + the actual
+headers on top of the actual transactions + evidence + last commit.
+
+## Proposed solution
+
+Therefore, we should
+
+1) Get rid of MaxTxs.
+2) Use MaxTxsBytes (rename to MaxDataBytes?) to limit the overall transactions size (Data).
+3) Introduce Max\* variables for header, evidence and last commit:
+  * MaxHeaderBytes
+  * MaxEvidenceBytes
+  * MaxLastCommitBytes
+4) Use these variables to limit the appropriate parts (header, evidence, ...)
+https://github.com/tendermint/tendermint/blob/8a1a79257e8e8b309cd35bb1fe40bf9b3330fd7d/types/block.go#L195
+
+Proposed default values are:
+
+- MaxHeaderBytes - 512 bytes (~200 bytes hashes + 200 bytes - 50 UTF-8 encoded symbols of chain ID)
+- MaxEvidenceBytes - 4096 bytes (10 DuplicateVoteEvidence ~350 bytes each)
+- MaxLastCommitBytes - 163840 (160 KB) (1000 votes ~150 bytes each)
```

Comment:
Not sure this one should be a parameter. Seems completely determined by number of validators and tendermint protocol right?

Reply:
Correct, but we still want to limit the block size. We know that TM currently does not support more than 300 validators (even less), so 1000 should be fine.

Reply:
Why not go the other way around?

i.e. Set a `MaxBlockSize`, and then derive tx size as follows:

```
MaxLastCommitBytes = function of number of validators currently enabled
MaxEvidenceBytes = function of consensus param: "maxNumEvidences"
MaxHeaderBytes = a consensus param
```
Then `MaxTxSize = MaxBlockSize - amino overhead - MaxLastCommitBytes - MaxEvidenceBytes - MaxHeaderBytes`. This makes MaxBlockSize the constant in consensus, not commit / tx size. (Max TxSize will decrease automatically as num validators increases, until block size is increased)

Reply:
> function of consensus param: "maxNumEvidences"

and `MaxEvidenceBytes` - max size of a single evidence

> function of number of validators currently enabled

and `MaxVoteBytes`, right?

> This makes MaxBlockSize the constant in consensus, not commit / tx size. (Max TxSize would then decrease automatically as num validators increases)

Right, so it's either a) we have a sum and block size is growing when number of validators increasing OR b) we have a fixed `BlockSize.MaxBytes` and number of txs is decreasing while number of validators increases

Reply:
I prefer the latter, fixed block size and space for txs decreases as validators increases. Fixing the block size rather than tx data size feels more natural to me, and allows for better predictions of storage requirements, hence my preference for it. 

MaxEvidenceBytes and MaxVoteBytes are calculated constants, not part of the config right? If so, then yes.